homeassistant/components/tts/__init__.py

```diff
@@ -0,0 +1,359 @@
+        provider.language = config.get(CONF_LANG)
+
+        cache_dir = yield from self.hass.loop.run_in_executor(
+            None, tts_cache_dir, self.hass, config.get(CONF_CACHE_DIR))
+
+        provider.cache_dir = cache_dir
+        self.providers[engine] = provider
+
+    @asyncio.coroutine
+    def async_get_url(self, engine, message):
+        """Get URL for play message.
+
+        This method is a coroutine.
+        """
+        provider = self.providers[engine]
+        key = hash((engine, message))
+
+        if provider.use_cache and key in self.speech_db:
+            file_name = self.speech_db[key][DB_FILE_NAME]
+        else:
+            file_name = yield from self.async_save_tts(engine, key, message)
+            if file_name is None:
+                return None
+
+        return "{}/{}/{}".format(
+            self.hass.config.api.base_url, engine, file_name)
+
+    @asyncio.coroutine
+    def async_save_tts(self, engine, key, message):
+        """Receive TTS and store for view.
+
+        This method is a coroutine.
+        """
+        provider = self.providers[engine]
+        data = yield from provider.async_run_tts(message)
+        file_name = "{}_{}.{}".format(hash(message), engine,
+                                      provider.file_format)
+
+        if data is None:
+            _LOGGER.error("No TTS from %s for '%s'", engine, message)
+            return None
+
+        def save_speech():
+            """Store speech to filesystem."""
+            with open(os.path.join(provider.cache_dir, file_name)) as speech:
+                speech.write(data)
+
+        try:
+            yield from self.hass.loop.run_in_executor(None, save_speech)
+        except OSError:
+            _LOGGER.error("Can't write %s", file_name)
+            return None
+
+        if provider.use_cache:
+            self.speech_db[key] = {
+                DB_MESSAGE: message,
+                DB_FILE_NAME: file_name,
+                DB_PROVIDER: engine,
+            }
+            self.hass.async_add_job(
+                self.async_update_db(key, self.speech_db[key]))
+
+        return file_name
+
+    @asyncio.coroutine
+    def async_update_db(self, key, data):
+        """Add device to YAML configuration file.
+
+        This method is a coroutine.
+        """
+        with (yield from self._is_updating):
+            yield from self.hass.loop.run_in_executor(
+                None, update_db, self.hass.config.path(YAML_TTS_DB),
+                key, data)
+
+
+class Provider(object):
+    """Represent a single provider."""
+
+    hass = None
+    use_cache = DEFAULT_CACHE
+    cache_dir = DEFAULT_CACHE_DIR
+    language = DEFAULT_LANG
+
+    @property
+    def file_format(self):
+        """Return file/audio format."""
+        raise NotImplementedError()
+
+    @property
+    def content_type(self):
+        """Return file/audio format."""
+        raise NotImplementedError()
+
+    def run_tts(self, message):
```

Comment:
let's rename it to `get_tts_audio`. I would expect this to return the content type too, maybe as a named tuple? That way a provider can return files in different formats.